### PR TITLE
Add LimitedOffsetsEnum to Limited offset token

### PR DESCRIPTION
**Pull request by luyuncheng**
In #67325 add `max_analyzed_offset` allows users to limit the highlighting of text fields.
but when using `"index_options": "offsets"` in mappings, this `max_analyzed_offset offset` can not be limited, 
because it would use `OffsetSource.ANALYSIS`and `PostingsOffsetStrategy` to highlight field.

When a string="Testing Fun Testing Fun" need to be highlight

Using `UnifiedHighlighter.OffsetSource.ANALYSIS` with `queryMaxAnalyzedOffset=10` would response: 
```
"Testing <b>Fun</b> Testing Fun"
```
because in #67325 add `LimitTokenOffsetAnalyzer`

__BUT USING__ `UnifiedHighlighter.OffsetSource.POSTINGS` which mapping is `index_options=offsets` and `queryMaxAnalyzedOffset=10` would response:
```
"Testing <b>Fun</b> Testing <b>Fun</b>"
```
because offset can not be limited by `LimitTokenOffsetAnalyzer`

may be can add a new `LimitedOffsetsEnum` to limit offset token

Link Issue: #86109

### CODE REVIEW

**Comment by rahuldimri** [comment]
is this issue available to work ?. or is there anything pending on this issue?

File: server/src/test/java/org/elasticsearch/lucene/search/uhighlight/CustomUnifiedHighlighterTests.java [hunk 1]
```diff
@@ -394,4 +396,102 @@ public void testExceedMaxAnalyzedOffset() throws Exception {
...
+            new String[] { "Testing <b>Fun</b> Testing Fun" },
+            29,
+            10,
+            UnifiedHighlighter.OffsetSource.POSTINGS
+        );
+    }
+
+
+    protected Analyzer wrapAnalyzer(Analyzer analyzer, Integer maxAnalyzedOffset) {
+        if (maxAnalyzedOffset != null) {
```
Comment (romseygeek) [inline server/src/test/java/org/elasticsearch/lucene/search/uhighlight/CustomUnifiedHighlighterTests.java:492]: Why do you need to wrap the analyzer here? Shouldn't it be handled by passing the max offset to CustomUnifiedHighlighter?

Comment (luyuncheng) [reply server/src/test/java/org/elasticsearch/lucene/search/uhighlight/CustomUnifiedHighlighterTests.java:492]: At beginning i want to describe how the analyzer runs with different analyzers(Default VS LimitTokenOffsetAnalyzer) works in different maxAnalyzedOffset. 
i think in this test case is unnecessary, but we can add random tests for different  maxAnalyzedOffset

Comment (romseygeek) [reply server/src/test/java/org/elasticsearch/lucene/search/uhighlight/CustomUnifiedHighlighterTests.java:492]: ++, random tests would be great

**Comment by romseygeek** [comment]
@elasticmachine test this please

**Comment by luyuncheng** [comment]
> ++, random tests would be great


@romseygeek  Thanks for the reviewing this.

At https://github.com/elastic/elasticsearch/pull/86110/commits/5d5e7018cb85dfee39f4a94844186f6319bf3e86 i removed the `wrapAnalyzer` and merge the assertHighlightOneDoc into one function
AND
i added a test for random different `OffsetSource` and `maxAnalyzedOffset `

File: server/src/test/java/org/elasticsearch/lucene/search/uhighlight/CustomUnifiedHighlighterTests.java [hunk 2]
```diff
@@ -394,4 +425,73 @@ public void testExceedMaxAnalyzedOffset() throws Exception {
...
+        );
+        assertHighlightOneDoc(
+            "text",
+            new String[] { "Testing Fun Testing Fun" },
+            analyzer,
+            query,
+            Locale.ROOT,
+            BreakIterator.getSentenceInstance(Locale.ROOT),
+            0,
+            // OLD Strategy would Response： "Testing <b>Fun</b> Testing <b>Fun</b>"
```
Comment (romseygeek) [inline server/src/test/java/org/elasticsearch/lucene/search/uhighlight/CustomUnifiedHighlighterTests.java:454]: This comment will be confusing in a couple of years time (what was the old strategy? why was it different?).  I think the current behaviour is a bug so it's fine to just test for it, no need to add a comment.

**Review by romseygeek — CHANGES_REQUESTED** [review]
Thanks @luyuncheng! I left a few small comments but I think this is close.

File: server/src/main/java/org/elasticsearch/lucene/search/uhighlight/LimitedOffsetsEnum.java [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+import org.apache.lucene.util.BytesRef;
+import org.elasticsearch.core.Nullable;
+
+import java.io.IOException;
+
+public class LimitedOffsetsEnum extends OffsetsEnum {
+    private final OffsetsEnum delegate;
+    private final Integer maxOffset;
+
+    public LimitedOffsetsEnum(OffsetsEnum delegate, @Nullable Integer maxOffset) {
```
[the remark this answers is not in the document]
Comment (luyuncheng) [reply server/src/main/java/org/elasticsearch/lucene/search/uhighlight/LimitedOffsetsEnum.java]: LGTM

File: server/src/test/java/org/elasticsearch/lucene/search/uhighlight/CustomUnifiedHighlighterTests.java (hunk 2 above)
Comment (luyuncheng) [reply server/src/test/java/org/elasticsearch/lucene/search/uhighlight/CustomUnifiedHighlighterTests.java]: LGTM

File: server/src/main/java/org/elasticsearch/lucene/search/uhighlight/LimitedOffsetsEnum.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (luyuncheng) [reply server/src/main/java/org/elasticsearch/lucene/search/uhighlight/LimitedOffsetsEnum.java]: In https://github.com/elastic/elasticsearch/pull/86110/commits/e18513f96d5edf5887329cf3b73927dd0587820a fixed it

**Comment by romseygeek** [comment]
@elasticmachine test this please

**Comment by romseygeek** [comment]
@elasticmachine generate changelog

File: docs/changelog/86110.yaml [hunk 4]
```diff
@@ -0,0 +1,6 @@
+pr: 86110
+summary: Add LimitedOffsetsEnum to Limited offset token
+area: Search/Highlighting
```
Comment (romseygeek) [inline docs/changelog/86110.yaml:3]: Can you change this to just `Search`? Apparently `Search/Highlighting` doesn't work here, 🤷

Comment (luyuncheng) [reply docs/changelog/86110.yaml:3]: OK

**Comment by luyuncheng** [comment]
@elasticmachine  test this please

**Comment by romseygeek** [comment]
@elasticmachine ok to test

**Comment by romseygeek** [comment]
Hi @luyuncheng can you run `./gradlew precommit` at the root and then push the relevant changes?

**Comment by romseygeek** [comment]
@elasticmachine update branch

**Review by romseygeek — APPROVED** [review]

**Comment by romseygeek** [comment]
Thanks for your patience @luyuncheng!

**Pull request merged**